src/diff/mod.rs

```diff
@@ -1,36 +1,285 @@
+    pub modified: Vec<ModifiedFile>,
+}
 
-pub struct Location {}
-
-pub enum Change {
-    AddLine {
-        path: NonEmpty<Label>,
-        location: Location,
-        contents: String,
-    },
-    RemoveLine {
-        path: NonEmpty<Label>,
-        location: Location,
-    },
-    MoveFile {
-        old_path: NonEmpty<Label>,
-        new_path: NonEmpty<Label>,
-    },
-    CreateFile {
-        path: NonEmpty<Label>,
-    },
-    DeleteFile {
-        path: NonEmpty<Label>,
-    },
-}
-
-pub struct Diff(pub Vec<Change>);
+pub struct CreateFile {
+    pub path: NonEmpty<Label>,
+}
+
+pub struct DeleteFile {
+    pub path: NonEmpty<Label>,
+}
+
+pub struct MoveFile {
+    pub old_path: NonEmpty<Label>,
+    pub new_path: NonEmpty<Label>,
+}
+
+pub struct ModifiedFile {
+    pub path: NonEmpty<Label>,
+    pub diff: FileDiff,
+}
+
+pub struct FileDiff {
+    // TODO
+}
 
 impl Diff {
-    // TODO(fintan): This is a bit more involved going to elide for now
-    #[allow(dead_code)]
-    fn diff<Repo>(_directory: Directory, _directory_: Directory) -> Self {
-        unimplemented!()
+    fn new() -> Self {
+        Diff {
+            created: Vec::new(),
+            deleted: Vec::new(),
+            moved: Vec::new(),
+            modified: Vec::new(),
+        }
+    }
+
+    // TODO: Direction of comparison is not obvious with this signature.
+    // For now using conventional approach with the right being "newer".
+    pub fn diff(left: Directory, right: Directory) -> Result<Diff, DiffError> {
+        let mut diff = Diff::new();
+        let path = Rc::new(RefCell::new(Vec::new()));
+        Diff::collect_diff(&left, &right, &path, &mut diff)?;
+
+        // TODO: Some of the deleted files may actually be moved (renamed) to one of the created files.
+        // Finding out which of the deleted files were deleted and which were moved will probably require
+        // performing some variant of the longest common substring algorithm for each pair in D x C.
+        // Final decision can be based on heuristics, e.g. the file can be considered moved,
+        // if len(LCS) > 0,25 * min(size(d), size(c)), and deleted otherwise.
+
+        Ok(diff)
+    }
+
+    fn collect_diff(old: &Directory, new: &Directory, parent_path: &Rc<RefCell<Vec<Label>>>, diff: &mut Diff)
+        -> Result<(), String> {
+
+        // TODO: Consider storing directory contents in sorted order
+        let old = get_sorted_contents(old);
+        let new = get_sorted_contents(new);
+
+        let mut old_iter = old.iter();
+        let mut new_iter = new.iter();
+        let mut old_entry_opt = old_iter.next();
+        let mut new_entry_opt = new_iter.next();
+
+        while old_entry_opt.is_some() || new_entry_opt.is_some() {
+            match (old_entry_opt, new_entry_opt) {
+                (Some(old_entry), Some(new_entry)) => {
+                    let cmp = new_entry.label().cmp(&old_entry.label());
+                    match cmp {
```

Comment:
This can just be `match new_entry.label().cmp(&old_entry.label())` :)

Reply:
Done